src/processor.rs

```diff
@@ -26,185 +26,275 @@ impl Processor {
 		    &[&[&b"escrow"[..], &[bump_seed]]], 
 		)?;
 
+		Self::close_pda_and_escrow(
+			pda_temp_token_account_info,
+			token_program_account_info,
+			initializers_main_account_info,
+			pda,
+			bump_seed,
+			pda_account_info,
+			escrow_account_info
+		)
+	}
+
+	/// Cancel can be called after init_escrow. If called after exchange, it's already too late
+	/// since exchange is atomic and tokens have been transferred, so there will be no effect.
+	/// Since tokens haven't actually been transferred from initializer main token account to
+	/// initializer temp token account, we don't need to actually transfer any tokens.
+	/// What we need to do is:
+	/// 1) Close PDA account
+	/// 2) Close escrow info
+	fn process_cancel(account_infos: &[AccountInfo], program_id: &Pubkey) -> ProgramResult {
+		let account_info_iter = &mut account_infos.iter();
+		let initializer_info = next_account_info(account_info_iter)?;
+
+		if !initializer_info.is_signer {
+			return Err(ProgramError::MissingRequiredSignature);
+		}
+
+		let initializer_token_account_info = next_account_info(account_info_iter)?;
+		if *initializer_token_account_info.owner != spl_token::id() {
+			return Err(ProgramError::IncorrectProgramId);
+		}
+
+		let escrow_account_info = next_account_info(account_info_iter)?;
+		let escrow = Escrow::unpack_unchecked(&escrow_account_info.try_borrow_data()?)?;
+		if !escrow.is_initialized() {
+			return Err(ProgramError::UninitializedAccount);
+		}
+
+		let token_program_account_info = next_account_info(account_info_iter)?;
+
+		let pda_temp_token_account_info = next_account_info(account_info_iter)?;
+		let (pda, bump_seed) = Pubkey::find_program_address(&[b"escrow"], program_id);
+
+		let initializers_main_account_info = next_account_info(account_info_iter)?;
+		// Validate Escrow matches instruction
+		if escrow.temp_token_account_pubkey != *pda_temp_token_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
+		if escrow.initializer_pubkey != *initializers_main_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
+		if escrow.initializer_token_to_receive_account_pubkey != *initializer_token_account_info.key {
+			return Err(ProgramError::InvalidAccountData);
+		}
+		
+		let pda_account_info = next_account_info(account_info_iter)?;
+
+		// Close initializer_token_account, return rent fees
+		**initializers_main_account_info.lamports.borrow_mut() = initializers_main_account_info.lamports()
+			.checked_add(initializer_token_account_info.lamports())
+			.ok_or(EscrowError::AmountOverflow)?;
+		**initializer_token_account_info.lamports.borrow_mut() = 0;
+		*initializer_token_account_info.try_borrow_mut_data()? = &mut [];
```

Comment:
and this is also repeated code with the account closing code in `close_pda_and_escrow`